ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp

```diff
@@ -0,0 +1,239 @@
+            {
+                auto C = feature_maps_shape[1];
+                auto feature_map_height = feature_maps_shape[2];
+                auto feature_map_width = feature_maps_shape[3];
+                auto num_rois = rois_shape[0];
+
+                CoordinateTransform feature_maps_transform(feature_maps_shape);
+                CoordinateTransform rois_transform(rois_shape);
+                CoordinateTransform out_transform(out_shape);
+
+                for (unsigned int roi_index = 0; roi_index < num_rois; roi_index++)
+                {
+                    // Get ROI`s corners
+                    T x1 = rois[rois_transform.index({roi_index, 0})] * spatial_scale;
+                    T y1 = rois[rois_transform.index({roi_index, 1})] * spatial_scale;
+                    T x2 = rois[rois_transform.index({roi_index, 2})] * spatial_scale;
+                    T y2 = rois[rois_transform.index({roi_index, 3})] * spatial_scale;
+
+                    T roi_width = std::max(x2 - x1, static_cast<T>(1.0));
+                    T roi_height = std::max(y2 - y1, static_cast<T>(1.0));
+
+                    T bin_width = roi_width / pooled_width;
+                    T bin_height = roi_height / pooled_height;
+
+                    uint64_t num_samples_in_bin = sampling_ratio * sampling_ratio;
+
+                    T sample_distance_x = bin_width / static_cast<T>(sampling_ratio);
+                    T sample_distance_y = bin_height / static_cast<T>(sampling_ratio);
+
+                    std::vector<std::pair<unsigned int, unsigned int>> pooling_points;
+                    std::vector<T> pooling_weights;
+
+                    pooling_points.reserve(4 * num_samples_in_bin * pooled_height * pooled_width);
+                    pooling_weights.reserve(4 * num_samples_in_bin * pooled_height * pooled_width);
+
+                    // Save the sample coords and weights as they will be identical across all
+                    // channels
+                    for (unsigned int y_bin_ind = 0; y_bin_ind < pooled_height; y_bin_ind++)
+                    {
+                        for (unsigned int x_bin_ind = 0; x_bin_ind < pooled_width; x_bin_ind++)
+                        {
+                            for (unsigned int y_sample_ind = 0; y_sample_ind < sampling_ratio;
+                                 y_sample_ind++)
+                            {
+                                T sample_y = y1 + static_cast<T>(y_bin_ind) * bin_height +
+                                             sample_distance_y * (static_cast<T>(y_sample_ind) +
+                                                                  static_cast<T>(0.5f));
+
+                                for (int64_t x_sample_ind = 0; x_sample_ind < sampling_ratio;
+                                     x_sample_ind++)
+                                {
+                                    T sample_x = x1 + static_cast<T>(x_bin_ind) * bin_width +
```

Comment:
Can you please indicate source of used formulas? In spec there is only "bilinear interpolation" keyword

Reply:
So the formulas for sample_x and sample_y are rather easy to figure out- we know that in each bin we need to take sampling_ratio * sampling_ratio samples, but how to figure out where they should be located? 
Well a few articles about ROI align described this in a similar fashion, but some differences:
-  for horizontal spacing just divide the bin width by sampling ratio, then space the first sample out by half the value of horizontal spacing - no link for this article as I don't have any on hand
-  for horizontal spacing just divide the bin width by sampling ratio+1, then space the first sample out by the value of horizontal spacing: [article](https://towardsdatascience.com/understanding-region-of-interest-part-2-roi-align-and-roi-warp-f795196fc193)

The rest of line 95 is just starting from ROI`s beginning and then shifting to appropriate bin.

Reply:
thanks, especially [article](https://towardsdatascience.com/understanding-region-of-interest-part-2-roi-align-and-roi-warp-f795196fc193) is very good ;)